packages/contracts-bedrock/src/L1/DataAvailabilityChallenge.sol

```diff
@@ -0,0 +1,343 @@
+        if (_challenge.challenger == address(0)) return ChallengeStatus.Uninitialized;
+
+        // if the challenge has a resolved block, it is resolved
+        if (_challenge.resolvedBlock != 0) return ChallengeStatus.Resolved;
+
+        // if the challenge's start block is in the resolve window, it is active
+        if (_isInResolveWindow(_challenge.startBlock)) return ChallengeStatus.Active;
+
+        // if the challenge's start block is not in the resolve window, it is expired
+        return ChallengeStatus.Expired;
+    }
+
+    /// @notice Challenge a data commitment at a given block number.
+    /// @dev The block number parameter is necessary for the contract to verify the challenge window,
+    ///      since the contract cannot access the block number of the commitment.
+    ///      The function reverts if the caller does not have a bond or if the challenge already exists.
+    /// @param challengedBlockNumber The block number at which the commitment was made.
+    /// @param challengedHash The data commitment that is being challenged.
+    function challenge(uint256 challengedBlockNumber, bytes32 challengedHash) external payable {
+        // deposit value sent with the transaction as bond
+        deposit();
+
+        // require the caller to have a bond
+        if (balances[msg.sender] < bondSize) {
+            revert BondTooLow(balances[msg.sender], bondSize);
+        }
+
+        // require the challenge status to be uninitialized
+        if (getChallengeStatus(challengedBlockNumber, challengedHash) != ChallengeStatus.Uninitialized) {
+            revert ChallengeExists();
+        }
+
+        // require the current block to be in the challenge window
+        if (!_isInChallengeWindow(challengedBlockNumber)) {
+            revert ChallengeWindowNotOpen();
+        }
+
+        // reduce the caller's balance
+        balances[msg.sender] -= bondSize;
+
+        // store the challenger's address, bond size, and start block of the challenge
+        challenges[challengedBlockNumber][challengedHash] =
+            Challenge({ challenger: msg.sender, lockedBond: bondSize, startBlock: block.number, resolvedBlock: 0 });
+
+        // emit an event to notify that the challenge status is now active
+        emit ChallengeStatusChanged(challengedHash, challengedBlockNumber, ChallengeStatus.Active);
+    }
+
+    /// @notice Resolve a challenge by providing the pre-image data of the challenged commitment.
+    /// @dev The provided pre-image data is hashed (keccak256) to verify that it matches the challenged commitment.
+    ///      The function reverts if the challenge is not active or if the resolve window is not open.
+    /// @param challengedBlockNumber The block number at which the commitment was made.
+    /// @param preImage The pre-image data corresponding to the challenged commitment.
+    function resolve(uint256 challengedBlockNumber, bytes32 challengedHash, bytes calldata preImage) external {
```

Comment:
How large can this pre-image be? If it's for a DA commitment it could be quite large and not fit in the calldata.

Reply:
we're limiting it to the max batch size to make sure it fits in one tx (this limitation is on the batcher side though)

Reply:
Can we add this to the specs + enforce it a protocol level?

Reply:
We need to have the datasize be limited inside the op-node and on the contracts. See also the slack thread for some other options.

I don't think that just setting this limit in the batcher is enough.